pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java

```diff
@@ -0,0 +1,177 @@
+  private final String _segmentName;
+  private final SegmentVersion _segmentVersion;
+  private final Set<ColumnMetadata> _textIndexColumns = new HashSet<>();
+
+  public TextIndexHandler(@Nonnull File indexDir, @Nonnull SegmentMetadataImpl segmentMetadata,
+      @Nonnull IndexLoadingConfig indexLoadingConfig, @Nonnull SegmentDirectory.Writer segmentWriter) {
+    _indexDir = indexDir;
+    _segmentWriter = segmentWriter;
+    _segmentName = segmentMetadata.getName();
+    _segmentVersion = SegmentVersion.valueOf(segmentMetadata.getVersion());
+
+    for (String column : indexLoadingConfig.getTextIndexColumns()) {
+      ColumnMetadata columnMetadata = segmentMetadata.getColumnMetadataFor(column);
+      if (columnMetadata != null) {
+        _textIndexColumns.add(columnMetadata);
+      }
+    }
+  }
+
+  /**
+   * Create text index for column during segment load time. Currently text index is always
+   * created (if enabled on a column) during segment generation (true for both offline
+   * and realtime segments). So this function is a NO-OP for case when a new segment is loaded
+   * after creation. However, when segment reload is issued in the following scenarios, we generate
+   * text index.
+   *
+   * SCENARIO 1: user enables text index on an existing column (table config change)
+   * SCENARIO 2: user adds a new column and enables text index (both schema and table config change)
+   *
+   * This function is a NO-OP for the above two cases. Later we can also add a segment generator
+   * config option to not necessarily generate text index during segment generation. When we do
+   * so, this function should be able to take care of that scenario too.
+   *
+   * For scenario 2, {@link org.apache.pinot.core.segment.index.loader.defaultcolumn.V3DefaultColumnHandler}
+   * would have already added the forward index for the column with default value. We use the forward
+   * index here to get the raw data and build text index.
+   *
+   * @throws IOException
+   */
+  public void createTextIndexesOnSegmentLoad()
+      throws Exception {
+    for (ColumnMetadata columnMetadata : _textIndexColumns) {
+      checkUnsupForEnablingTextIndexOnExistingColumn(columnMetadata);
+      createTextIndexForColumn(columnMetadata);
+    }
+  }
+
+  /**
+   * Right now the text index is supported on RAW (non-dictionary encoded)
+   * single-value STRING columns. Eventually we will relax the constraints
+   * step by step.
+   * For example, later on user should be able to create text index on
+   * a dictionary encoded STRING column that also has native Pinot's inverted
+   * index. We can also support it on BYTE columns later.
+   * @param columnMetadata metadata for column
+   */
+  private void checkUnsupForEnablingTextIndexOnExistingColumn(ColumnMetadata columnMetadata) {
+    if (columnMetadata.hasDictionary()) {
+      throw new UnsupportedOperationException("Text index is currently not supported on dictionary encoded columns");
```

Comment:
Same comment as before, about column names and use of Precondition with a boolean return

Reply:
done